sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala

```diff
@@ -112,11 +112,24 @@ private[hive] object IsolatedClientLoader extends Logging {
       hadoopVersion: String,
       ivyPath: Option[String],
       remoteRepos: String): Seq[URL] = {
+    val hadoopJarNames = if (hadoopVersion.startsWith("3")) {
```

Comment:
Will this break if `hadoopVersion` is 3.1.0, 3.2.1, etc. (due to the previous issues with the shaded client JARs)?

Reply:
Do you mean [HADOOP-16080](https://issues.apache.org/jira/browse/HADOOP-16080)? yes things could still break in the following cases.

1. users build Spark without `-Phadoop-cloud` AND use a version doesn't have the fix in HADOOP-16080, such as:
```
$ bin/spark-shell --packages org.apache.hadoop:hadoop-aws:3.2.0,org.apache.hadoop:hadoop-common:3.2.0
```
However I think we should recommend users to stick to the same version used by Spark, i.e., 3.2.2

2. users build Spark with custom Hadoop version such as 3.1.0/3.2.1 you mentioned via the `hadoop.version` property, and use this to talk to cloud storage like S3. 

To enable these use cases we may have to introduce another Maven property to switch back to non-shaded client, and update here as well.

Reply:
Yah I think (2) is my primary concern. We support building against a custom version using `-Dhadoop.version`, but right now it will break if you use `-Phadoop-3 -Dhadoop.version=3.1.0`. This one I believe you can at least work around by changing the `-Dhadoop-client-{runtime,api,minicluster}.artifact` properties, but here there's no way to work around it.

Reply:
Got it. Yes we can be more precise here, such as using a version map. I can update this.